### PR TITLE
feat: add Pebble check-failed and check-recovered events

### PR DESCRIPTION
**Pull request by tonyandrewmeyer**
This creates new `<container-name>-pebble-check-failed` and `<container-name>-pebble-check-failed` hooks, which are triggered by Pebble change-update notices of particular kinds & statuses, with the effect that we'll now respond to Pebble checks reaching the failure threshold, or passing after having reached the threshold, by sending a pebble-check-failed or pebble-check-recovered event to the charm.

### Test the hooks directly:
```shell
# Pack and deploy the test charm
$ juju add-model t
$ cd testcharms/charms/pebble-checks
$ charmcraft pack
$ juju deploy ./juju-qa-pebble-checks_ubuntu-22.04-amd64.charm --resource ubuntu-image=public.ecr.aws/ubuntu/ubuntu:22.04

# The check is set to succeed if `/trigger/` exists, so make it.
$ juju ssh --container ubuntu juju-qa-pebble-checks/0 mkdir /trigger/

$ juju status  # unit status should settle to "active" after pebble-ready

# Make the check fail
$ juju ssh --container ubuntu juju-qa-pebble-checks/0 rmdir /trigger/
$ juju status  # will now say status "maintenance" with message "check failed: exec-check"

# Make the check pass again
$ juju ssh --container ubuntu juju-qa-pebble-checks/0 mkdir /trigger/
$ juju status  # will now say status "active" with message "check recovered: exec-check"
```

### Test the shell test:
`./main.sh -p k8s -c minikube sidecar test_pebble_checks`

## Documentation changes

 * [CHARMTECH-160](https://warthogs.atlassian.net/browse/CHARMTECH-160), [CHARMTECH-161](https://warthogs.atlassian.net/browse/CHARMTECH-161), [CHARMTECH-162](https://warthogs.atlassian.net/browse/CHARMTECH-162) cover updating the juju.is (main and SDK) docs

## Links

* [OP046](https://docs.google.com/document/d/13ItH8l5ZSmmv9WqZXpqjV8EifZU5e3zxINiNLubnC68/edit)
* [juju/charm PR](https://github.com/juju/charm/pull/432)
* [ops PR](https://github.com/canonical/operator/pull/1281)
* [pebble PR](https://github.com/canonical/pebble/pull/444)
* [jhack PR](https://github.com/canonical/jhack/pull/172)
* [Scenario PR](https://github.com/canonical/ops-scenario/pull/151)

**Jira card:** [CHARMTECH-109](https://warthogs.atlassian.net/browse/CHARMTECH-109)

[CHARMTECH-160]: https://warthogs.atlassian.net/browse/CHARMTECH-160?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: worker/uniter/pebblenotices.go [hunk 1]
```diff
@@ -131,11 +131,59 @@ func (n *pebbleNoticer) run(containerName string) (err error) {
...
+		}
+	case client.ChangeUpdateNotice:
+		data := notice.LastData
+		// Since the charm is triggering most Pebble changes, there is little
+		// value in feeding events based on those changes back to the charm.
+		// As such, we do not send events for all change-updated notices (see
+		// OP045 for more background). However, perform-check and recover-check
+		// change kinds reflect changes in the workload state, so are useful to
+		// charms.
+		if data["kind"] != "perform-check" && data["kind"] != "recover-check" {
```
Comment (benhoyt) [inline worker/uniter/pebblenotices.go:155]: You could pull out `kind := data["kind"]` above as a named variable, seeing you're using it 6x here.

File: worker/uniter/runner/context/context.go [hunk 2]
```diff
@@ -1882,6 +1888,30 @@ func (ctx *HookContext) WorkloadName() (string, error) {
 	return ctx.workloadName, nil
 }
 
+// WorkloadNoticeType returns the type of the notice for workload notice hooks.
+func (ctx *HookContext) WorkloadNoticeType() (string, error) {
```
Comment (benhoyt) [inline worker/uniter/runner/context/context.go:1892]: It looks like these new functions are only used by tests. Do they need to live here?

**Review by benhoyt — COMMENTED** [review]
Nice work! Couple of minor comments. But obviously the Juju team will have more to say.

File: worker/uniter/runner/context/context.go (hunk 2 above)
Comment (tonyandrewmeyer) [reply worker/uniter/runner/context/context.go:1892]: I was copying where the other similar fields were exposed, but, no, they aren't used outside of the tests so could be moved (into one of the _test files I assume?). I might need some help with that, though, I'm not sure how to provide access in contextfactory_test to the fields that are in context. It's the same package but I get errors if I try to access it from contextfactory. I can try to figure it out Monday, but might need some hints.

Comment (benhoyt) [reply worker/uniter/runner/context/context.go:1892]: Okay, I see -- if the other similar fields are exposed this way, that's probably fine. Yeah, I see some of the others exposed in a similar way are only used by tests. The alternative would be to put them on `HookContext` in `export_test.go` -- which is in `package context` but has a `_test.go` filename so is compiled only when compiling tests -- see `HookContext.PendingSecretRemoves`, for example. But in this case I think you're right, precedent says to put them right where you have them.

**Review by hpidcock — APPROVED** [review]
Awesome work. Testing passes. Thanks for adding a shell test.

**Pull request merged**